server/channels/app/post.go

```diff
@@ -773,51 +773,65 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
 			return false, nil
 		}
-		return false, err
-	}
-
-	channelMembers, err := a.GetChannelMembersPage(c, post.ChannelId, 0, 10000000)
-	if err != nil {
-		return false, err
+		return false, appErr
 	}
 
-	permalinkPreviewedChannel, err := a.GetChannel(c, previewedPost.ChannelId)
-	if err != nil {
-		if err.StatusCode == http.StatusNotFound {
+	permalinkPreviewedChannel, appErr := a.GetChannel(c, previewedPost.ChannelId)
+	if appErr != nil {
+		if appErr.StatusCode == http.StatusNotFound {
 			mlog.Warn("channel containing permalinked post not found", mlog.String("referenced_channel_id", previewedPost.ChannelId))
 			return false, nil
 		}
-		return false, err
+		return false, appErr
 	}
 
 	permalinkPreviewedPost := post.GetPreviewPost()
-	for _, cm := range channelMembers {
-		if permalinkPreviewedPost != nil {
-			post.Metadata.Embeds[0].Data = permalinkPreviewedPost
-		}
+	if permalinkPreviewedPost != nil {
+		post.Metadata.Embeds[0].Data = permalinkPreviewedPost
+	} else {
+		return false, nil
+	}
 
-		postForUser := a.sanitizePostMetadataForUserAndChannel(c, post, permalinkPreviewedPost, permalinkPreviewedChannel, cm.UserId)
+	// This is a hack to cache the JSON representation of the post metadata
+	// to be attached to the websocket event if the user doesn't have permission.
+	// We wipe the data, get the JSON, and restore the data again.
+	// This simplifies the data handling in the channel hook, where otherwise
+	// multiple goroutines would be racing to modify the Metadata field.
+	// And deep-copying the Embed data also has challenges because the type is "any".
+	post.Metadata.Embeds[0].Data = nil
+	modifiedJSON, err := post.ToJSON()
+	if err != nil {
+		mlog.Warn("Failed to encode post to JSON", mlog.Err(err))
+	}
+	post.Metadata.Embeds[0].Data = permalinkPreviewedPost
 
-		// Using DeepCopy here to avoid a race condition
-		// between publishing the event and setting the "post" data value below.
-		messageCopy := message.DeepCopy()
-		broadcastCopy := messageCopy.GetBroadcast()
-		broadcastCopy.UserId = cm.UserId
-		messageCopy.SetBroadcast(broadcastCopy)
+	// Modifying post for the original sender, in case they don't have permission.
+	// This wipes off the original post content. And then the marshalled post content
+	// in the websocket event is taken care of in the channel hook.
+	if permalinkPreviewedChannel != nil && !a.HasPermissionToReadChannel(c, post.UserId, permalinkPreviewedChannel) {
+		post.Metadata.Embeds[0].Data = nil
+	}
 
-		postJSON, jsonErr := postForUser.ToJSON()
-		if jsonErr != nil {
-			mlog.Warn("Failed to encode post to JSON", mlog.Err(jsonErr))
+	broadcastCopy := message.GetBroadcast()
+	broadcastCopy.ChannelHook = func(userID string, ev *model.WebSocketEvent) bool {
+		hasChange := a.shouldSanitizePostMetadataForUserAndChannel(c, post, permalinkPreviewedChannel, userID)
+		if !hasChange {
+			// If there is no change, then the original post which was attached
+			// will get sent.
```

Comment:
I stumbled over this again, thinking the "original post" was the post set by the caller, and wondering where the permalink preview got set, before remembering that we (sometimes) set it above, so it's not so "original" anymore as the comment suggests. More and more, I'm wondering if we should bite the bullet and just fix this with a proper `publishWebsocketEventForPost`. It's extremely hard to reason about the correctness of the existing code + changes.

Reply:
The only thing that I am wondering is that, in that case we'll be doing `ev.Add()` inside the method but there'll be a lot of other `ev.Add()` outside the method as well for other stuff. For readers going through the higher layer, they'll wonder "where is the `post` attribute getting set?" only to find out there's this one method which sets it.